### PR TITLE
[ci] fix: run wan e2e in bf16

### PR DESCRIPTION
**Pull request by Luosuu**
### What does this PR do?

Run the Wan DiT e2e alignment test in bf16 mixed precision on the production FA2 path instead of the temporary fp32/eager workaround added during the PyTorch 2.11 / CUDA 13 upgrade.

This keeps Wan e2e coverage on explicit bf16 FSDP mixed precision:

- `--train.accelerator.fsdp_config.mixed_precision.enable=True`
- `--train.accelerator.fsdp_config.mixed_precision.param_dtype=bfloat16`
- `--train.accelerator.fsdp_config.mixed_precision.cast_forward_inputs=True`

### Design & Code Changes

- Replace Wan-specific fp32 FSDP args in `tests/e2e/utils.py` with explicit bf16 mixed-precision args.
- Pin Wan GPU e2e attention to `flash_attention_2`; keep Wan RoPE and RMSNorm eager on GPU so this CI test isolates the production FA2 attention path from unrelated fused-kernel numerics.
- Keep Wan e2e on the same controlled finite fixture shape as the Wan forward parity test: zero hidden/text/latent inputs, random nonzero training targets at `1e-2` scale, stable norm parameters, and nonzero constant non-norm checkpoint parameters.
- Tighten Wan e2e loss/mse absolute tolerance from `1e-5` to `1e-8` while keeping `rtol=1e-5`.
- Add per-metric comparison tolerances so only Wan `grad_norm` gets a targeted `2e-7` absolute tolerance for NPU reduction drift; loss/mse stay at `1e-8`.
- Add a command-generation regression test so Wan e2e cannot silently drift back to fp32-only or eager-attention coverage.
- Keep ordinary single-rank Wan `flash_attention_2` on diffusers native flash backend for pristine diffusers parity.
- Route T2V `veomni_flash_attention_2_with_sp` through VeOmni Transformers FA2 wrapper after explicit Ulysses gather/scatter.
- Apply Wan T2V Ulysses all-to-all before RoPE on the SP FA2 path so Q/K use full un-sliced rotary positions, matching the Qwen Image SP ordering.
- Add guards for unsupported Wan SP layouts: CP is rejected, Wan SP requires `veomni_flash_attention_2_with_sp`, hidden/rotary sequence lengths must match and divide SP size, attention heads must divide Ulysses size, and post-gather Q/K/V sequence lengths must match.
- Keep I2V added-KV attention on diffusers flash to avoid changing image-context layout.
- Add Wan bf16 forward coverage, because Wan is diffusers-backed and not covered by the generic transformers logits/forward parity tests:
  - single-rank pristine diffusers parity for both `flash_attention_2` and `veomni_flash_attention_2_with_sp` when the pristine reference is finite;
  - SP=2 forward coverage requiring finite output and exact equality to a no-SP Wan FA2 baseline when that baseline is itself runnable and finite.
- Add `tests/models/test_wan_diffusers_forward_equal.py` to `gpu_unit_tests_v5`.
- Make the shared HF FA wrapper pass contiguous Q/K/V tensors into `_flash_attention_forward`.

### Test

- `ruff format tests/e2e/test_e2e_parallel.py tests/tools/comparison_utils.py veomni/data/dummy_dataset.py`
- `ruff check tests/e2e/test_e2e_parallel.py tests/tools/comparison_utils.py veomni/data/dummy_dataset.py tests/models/test_wan_diffusers_forward_equal.py veomni/models/diffusers/wan_t2v/wan_transformer/modeling_wan_transformer.py tests/tools/training_utils.py tests/e2e/utils.py`
- `python -m py_compile tests/e2e/test_e2e_parallel.py tests/tools/comparison_utils.py veomni/data/dummy_dataset.py tests/models/test_wan_diffusers_forward_equal.py veomni/models/diffusers/wan_t2v/wan_transformer/modeling_wan_transformer.py tests/tools/training_utils.py tests/e2e/utils.py`
- `python tests/special_sanity/check_device_api_usage.py -d tests/e2e`
- `python tests/special_sanity/check_device_api_usage.py -d tests/tools`
- `python tests/special_sanity/check_device_api_usage.py -d veomni/data`
- `PYTHONPATH=. pytest -q tests/e2e/test_e2e_parallel.py::test_wan_dit_uses_bfloat16_mixed_precision_and_flash_attention`
  - Local result: command-generation test passes; full Wan GPU e2e cannot run locally because torch falls back to CPU due the local CUDA driver/toolkit mismatch.
- B200 worker attempt: not useful for this repo env because `.venv` torch still detects CUDA driver `12090` and falls back to CPU, failing distributed backend setup before model execution. L20 GitHub CI remains decisive.
- ModelChef review gate: final verdict `safe`.

### CI Notes

- GPU unit has repeatedly passed the new Wan diffusers/VeOmni bf16 FA2 forward parity coverage.
- NPU e2e passed after the targeted grad-norm tolerance while keeping Wan loss/mse strict.
- `dba1301b` keeps FA2/bf16, preserves norm scales in the toy checkpoint, uses a controlled finite e2e fixture, keeps strict `grad_norm` comparison with a targeted NPU tolerance, and tightens Wan loss/mse absolute tolerance.

### CODE REVIEW

**Comment by Luosuu** [comment]
Update after debugging the PyTorch 2.11 Wan bf16 NaN:

- Root cause: Wan T2V `bf16` grad-enabled FA2 self-attention hit non-finite outputs on the dense FA2 path under PyTorch 2.11/L20. The first non-finite module was `blocks.0.attn1`, before loss/backward/optimizer/FSDP/SP.
- Fix: keep the public `flash_attention_2` path, but route Wan T2V FA2 self-attention through flash-attn varlen/unpad with a full-ones mask, preserving attention semantics and avoiding the dense-path NaN. This is not a fp32/eager workaround.
- Follow-up fix: the Wan SP guard now validates the active attention processor implementation instead of relying on `config._attn_implementation`, which was not reliable in the CI construction path.
- Coverage added/updated: Wan diffusers-vs-VeOmni forward parity, FA2 backward finite check, nested dtype casting, and SP=2 vs no-SP FA2 parity.

Current head `a8ef3f5c` is green on GPU unit, GPU e2e, NPU unit, NPU e2e, lint, title, docs, device-api, and CLA. The remaining merge blocker is reviewer approval (`reviewDecision=REVIEW_REQUIRED`).

**Comment by Luosuu** [comment]
Update for current head `56ab2de4`:

- Root cause: Wan T2V `bf16` grad-enabled FA2 self-attention hit non-finite outputs on the dense FA2 path under PyTorch 2.11/L20. The first non-finite module was `blocks.0.attn1`, before loss/backward/optimizer/FSDP/SP.
- Final fix: keep the public `flash_attention_2` path and route Wan FA2 through the existing shared VeOmni/HF attention entry. Since DiT trainer feeds full per-sample sequences, `WanSPAttnProcessor` now passes explicit `cu_seq_lens_q/k` and `max_length_q/k` built from the full Q/K sequence lengths, so HF selects its varlen FA2 path without a synthetic mask or a Wan-private `_flash_attention_forward` helper.
- SP handling: Wan forward computes full hidden states and rotary embeddings, then slices `hidden_states` and `rotary_emb` together before transformer blocks. This removes the earlier `use_sp_before_rotary` special case; the processor applies rotary on aligned local slices and then uses the normal SP gather before FA2.
- Coverage added/updated: Wan diffusers-vs-VeOmni forward parity, FA2 backward finite check, nested dtype casting, SP=2 vs no-SP FA2 parity, and the Wan bf16 e2e command-generation check. Gemini review threads are resolved/outdated.

All checks are green on the current head: GPU unit, GPU e2e, NPU unit, NPU e2e, lint, title, docs, device-api, and CLA. The remaining merge blocker is reviewer approval (`reviewDecision=REVIEW_REQUIRED`).

**Comment by Luosuu** [comment]
All checks are green on current head `56ab2de4`; unresolved review threads are resolved/outdated. The only remaining blocker is required reviewer approval. cc @TimYangst @Coach257 @FoolPlayer

File: tests/e2e/utils.py [hunk 1]
```diff
@@ -82,9 +82,8 @@ def __str__(self):
 
 _SP_SIZE = [1, 2]
 _EP_SIZE = [1, 2]
-_WAN_FLOAT32_FSDP_ARGS = [
-    "--train.accelerator.fsdp_config.mixed_precision.param_dtype=float32",
-    "--train.accelerator.fsdp_config.mixed_precision.cast_forward_inputs=False",
+_WAN_BFLOAT16_TRAINING_ARGS = [
+    "--train.accelerator.fsdp_config.mixed_precision.enable=False",
 ]
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply tests/e2e/utils.py:89]: Agree, we need to test mixed precison e2e.

**Review by Coach257 — APPROVED** [review]

**Review by FoolPlayer — APPROVED** [review]
